File: Source/programmers/SWI.cs

namespace FlashcatUSB {
    public class SWI_Programmer : FlashcatUSB.MemoryDeviceUSB {
        private FlashcatUSB.USB.FCUSB_DEVICE FCUSB;

        public event PrintConsoleEventHandler PrintConsole;

        public delegate void PrintConsoleEventHandler(string message);

        public event SetProgressEventHandler SetProgress;

        public delegate void SetProgressEventHandler(int percent);

        private string SWI_DEV_NAME;
        private int SWI_DEV_SIZE; // Total bytes of the device 
        private ushort SWI_DEV_PAGE;

        public SWI_Programmer(FlashcatUSB.USB.FCUSB_DEVICE parent_if) {
            FCUSB = parent_if;
        }

        public bool DeviceInit() {
            var chip_id = new byte[3];
            bool detect_result = FCUSB.USB_CONTROL_MSG_IN(FlashcatUSB.USB.USBREQ.SWI_DETECT, ref chip_id, (uint)FlashcatUSB.MainApp.MySettings.SWI_ADDRESS);
            uint SWI_ID_DATA = (uint)(chip_id[0] << 16 | chip_id[1] << 8 | chip_id[2]);
            switch (SWI_ID_DATA) {
                case 0xD200U: {
                        SWI_DEV_NAME = "Microchip AT21CS01";
                        SWI_DEV_SIZE = 128;
                        SWI_DEV_PAGE = 8;
                        break;
                    }

                case 0xD380U: {
                        SWI_DEV_NAME = "Microchip AT21CS11";
                        SWI_DEV_SIZE = 128;
                        SWI_DEV_PAGE = 8;
                        break;
                    }

                default: {
                        return false;
                    }
            }

            return true;
        }

        public string DeviceName {
            get {
                return SWI_DEV_NAME;
            }
        }

        public long DeviceSize {
            get {
                return SWI_DEV_SIZE;
            }
        }

        public uint SectorSize(uint sector, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            return (uint)DeviceSize;
        }

        public byte[] ReadData(long flash_offset, long data_count, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            var setup_data = GetSetupPacket((uint)flash_offset, (uint)data_count, SWI_DEV_PAGE);
            var data_out = new byte[(int)(data_count - 1L + 1)];
            bool result = FCUSB.USB_SETUP_BULKIN(FlashcatUSB.USB.USBREQ.SWI_READ, setup_data, ref data_out, 0U);
            if (!result)
                return null;
            return data_out;
        }

        public bool WriteData(long flash_offset, byte[] data_to_write, FlashcatUSB.WriteParameters Params = null) {
            var setup_data = GetSetupPacket((uint)flash_offset, (uint)data_to_write.Length, SWI_DEV_PAGE);
            bool result = FCUSB.USB_SETUP_BULKOUT(FlashcatUSB.USB.USBREQ.SWI_WRITE, setup_data, data_to_write, 0U);
            return result;
        }

        public bool EraseDevice() {
            return true; // EEPROM does not support erase commands
        }

        public void WaitUntilReady() {
            FlashcatUSB.Utilities.Main.Sleep(10);
        }

        public long SectorFind(uint SectorIndex, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            return 0L;
        }

        public bool SectorErase(uint SectorIndex, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            return true;
        }

        public uint SectorCount() {
            return 1U;
        }

        public bool SectorWrite(uint SectorIndex, byte[] data, FlashcatUSB.WriteParameters Params = null) {
            return false; // Not supported
        }

        private byte[] GetSetupPacket(uint Address, uint Count, ushort PageSize) {
            byte addr_bytes = 0;
            var data_in = new byte[11];
            data_in[0] = (byte)(Address & 255L);
            data_in[1] = (byte)(Address >> 8 & 255L);
            data_in[2] = (byte)(Address >> 16 & 255L);
            data_in[3] = (byte)(Address >> 24 & 255L);
            data_in[4] = (byte)(Count & 255L);
            data_in[5] = (byte)(Count >> 8 & 255L);
            data_in[6] = (byte)(Count >> 16 & 255L);
            data_in[7] = (byte)(Count >> 24 & 255L);
            data_in[8] = (byte)(PageSize & 255); // This is how many bytes to increment between operations
            data_in[9] = (byte)(PageSize >> 8 & 255);
            data_in[10] = 1;
            return data_in;
        }
    }
}

File: Source/programmers/SWI.vb
Imports FlashcatUSB.FlashMemory

Public Class SWI_Programmer : Implements MemoryDeviceUSB
    Private FCUSB As USB.FCUSB_DEVICE

    Public Event PrintConsole(message As String) Implements MemoryDeviceUSB.PrintConsole
    Public Event SetProgress(percent As Integer) Implements MemoryDeviceUSB.SetProgress

    Public MyFlashDevice As SWI

    Sub New(parent_if As USB.FCUSB_DEVICE)
        FCUSB = parent_if
    End Sub

    Public Function DeviceInit() As Boolean Implements MemoryDeviceUSB.DeviceInit
        Dim chip_id(2) As Byte
        Dim detect_result As Boolean = FCUSB.USB_CONTROL_MSG_IN(USB.USBREQ.SWI_DETECT, chip_id, MySettings.SWI_ADDRESS)
        Dim SWI_ID_DATA As UInt32 = (CUInt(chip_id(0)) << 16) Or (CUInt(chip_id(1)) << 8) Or CUInt(chip_id(2))
        Dim MFG_CODE As Byte = CByte(CUInt(SWI_ID_DATA >> 12) And &HF)
        Dim PART As UInt16 = CUShort(SWI_ID_DATA >> 3) And &H1FFUS
        If MFG_CODE = &HD Then Return False
        If PART = &H40 Then
            MyFlashDevice = New SWI("Microchip AT21CS01", &HD, &H40, 128, 8)
        ElseIf PART = &H70 Then
            MyFlashDevice = New SWI("Microchip AT21CS11", &HD, &H70, 128, 8)
        Else
            Return False
        End If
        Return True
    End Function

    Public ReadOnly Property GetDevice As Device Implements MemoryDeviceUSB.GetDevice
        Get
            Return Me.MyFlashDevice
        End Get
    End Property

    Public ReadOnly Property DeviceName As String Implements MemoryDeviceUSB.DeviceName
        Get
            Return Me.MyFlashDevice.NAME
        End Get
    End Property

    Public ReadOnly Property DeviceSize As Long Implements MemoryDeviceUSB.DeviceSize
        Get
            Return Me.MyFlashDevice.FLASH_SIZE
        End Get
    End Property

    Public Function SectorSize(sector As Integer) As Integer Implements MemoryDeviceUSB.SectorSize
        Return CInt(Me.DeviceSize)
    End Function

    Public Function ReadData(flash_offset As Long, data_count As Integer) As Byte() Implements MemoryDeviceUSB.ReadData
        Dim setup_data() As Byte = GetSetupPacket(CUInt(flash_offset), CUInt(data_count), CUShort(Me.MyFlashDevice.PAGE_SIZE))
        Dim data_out(CInt(data_count) - 1) As Byte
        Dim result As Boolean = FCUSB.USB_SETUP_BULKIN(USB.USBREQ.SWI_READ, setup_data, data_out, 0)
        If Not result Then Return Nothing
        Return data_out
    End Function

    Public Function WriteData(flash_offset As Long, data_to_write() As Byte, Optional Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.WriteData
        Dim setup_data() As Byte = GetSetupPacket(CUInt(flash_offset), CUInt(data_to_write.Length), CUShort(Me.MyFlashDevice.PAGE_SIZE))
        Dim result As Boolean = FCUSB.USB_SETUP_BULKOUT(USB.USBREQ.SWI_WRITE, setup_data, data_to_write, 0)
        Return result
    End Function

    Public Function EraseDevice() As Boolean Implements MemoryDeviceUSB.EraseDevice
        Return True 'EEPROM does not support erase commands
    End Function

    Public Sub WaitUntilReady() Implements MemoryDeviceUSB.WaitUntilReady
        Utilities.Sleep(10)
    End Sub

    Public Function SectorFind(SectorIndex As Integer) As Long Implements MemoryDeviceUSB.SectorFind
        Return 0
    End Function

    Public Function SectorErase(SectorIndex As Integer) As Boolean Implements MemoryDeviceUSB.SectorErase
        Return True
    End Function

    Public Function SectorCount() As Integer Implements MemoryDeviceUSB.SectorCount
        Return 1
    End Function

    Public Function SectorWrite(SectorIndex As Integer, data() As Byte, Optional Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.SectorWrite
        Return False 'Not supported
    End Function

    Private Function GetSetupPacket(Address As UInt32, Count As UInt32, PageSize As UInt16) As Byte()
        Dim addr_bytes As Byte = 0
        Dim data_in(10) As Byte
        data_in(0) = CByte(Address And 255)
        data_in(1) = CByte((Address >> 8) And 255)
        data_in(2) = CByte((Address >> 16) And 255)
        data_in(3) = CByte((Address >> 24) And 255)
        data_in(4) = CByte(Count And 255)
        data_in(5) = CByte((Count >> 8) And 255)
        data_in(6) = CByte((Count >> 16) And 255)
        data_in(7) = CByte((Count >> 24) And 255)
        data_in(8) = CByte(PageSize And 255) 'This is how many bytes to increment between operations
        data_in(9) = CByte((PageSize >> 8) And 255)
        data_in(10) = 1
        Return data_in
    End Function

End Class
